numbat/src/command.rs

```diff
@@ -0,0 +1,472 @@
+    /// which contains the input, and a `code_source_id`, for reporting errors
+    pub fn new(inner: SourcelessCommandParser<'a>, code_source_id: usize) -> Self {
+        Self {
+            inner,
+            code_source_id,
+        }
+    }
+
+    pub fn set_code_source_id(&mut self, code_source_id: usize) {
+        self.code_source_id = code_source_id;
+    }
+
+    /// Get the span starting at the start of the word at `word_index`, through the end of
+    /// the last word represented by `word_boundaries`
+    ///
+    /// ## Panics
+    /// If `word_index` is out of bounds, ie `word_index >= word_boundaries.len()`
+    fn span_through_end(&self, word_index: usize) -> Span {
+        let start = self.inner.word_boundaries[word_index].0;
+        let end = self.inner.word_boundaries.last().unwrap().1;
+        self.span_from_boundary((start, end))
+    }
+
+    /// Get the span between indices given by `start` and `end`
+    ///
+    /// The only role of `&self` here is to provide the `code_source_id`
+    fn span_from_boundary(&self, (start, end): (u32, u32)) -> Span {
+        Span {
+            start: SourceCodePositition {
+                byte: start,
+                line: 1,
+                position: start,
+            },
+            end: SourceCodePositition {
+                byte: end,
+                line: 1,
+                position: end,
+            },
+            code_source_id: self.code_source_id,
+        }
+    }
+
+    fn err_at_idx(&self, index: usize, err_msg: &'static str) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidCommand(err_msg),
+            span: self.span_from_boundary(self.inner.word_boundaries[index]),
+        }
+    }
+
+    fn err_through_end_from(&self, index: usize, err_msg: &'static str) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidCommand(err_msg),
+            span: self.span_through_end(index),
+        }
+    }
+
+    /// Attempt to parse the input provided to Self::new as a command, such as "help",
+    /// "list <args>", "quit", etc
+    ///
+    /// Returns:
+    /// - `Ok(Command)`, if the input is a valid command with correct arguments
+    /// - `Err(ParseError)`, if the input starts with a valid command but has the wrong
+    ///   number or kind of arguments, e.g. `list foobar`
+    pub fn parse_command(&mut self) -> Result<Command, ParseError> {
+        macro_rules! ensure_zero_args {
+            ($parser:expr, $cmd:literal) => {{
+                ensure_zero_args!($parser, $cmd, "")
+            }};
+            ($parser:expr, $cmd: literal, $err_msg_suffix:literal) => {{
+                if $parser.inner.words.next().is_some() {
+                    return Err($parser.err_through_end_from(
+                        1,
+                        concat!("`", $cmd, "` takes 0 arguments", $err_msg_suffix),
+                    ));
+                }
+            }};
+        }
```

Comment:
Do we really need a macro here? If we are okay with a few allocations (this is just the error case, so it won't affect performance), we could do something like this:

```patch
diff --git a/numbat/src/command.rs b/numbat/src/command.rs
index 85070d3..b9da58d 100644
--- a/numbat/src/command.rs
+++ b/numbat/src/command.rs
@@ -137,20 +137,32 @@ impl<'a> CommandParser<'a> {
         }
     }
 
-    fn err_at_idx(&self, index: usize, err_msg: &'static str) -> ParseError {
+    fn err_at_idx(&self, index: usize, err_msg: impl Into<String>) -> ParseError {
         ParseError {
-            kind: ParseErrorKind::InvalidCommand(err_msg),
+            kind: ParseErrorKind::InvalidCommand(err_msg.into()),
             span: self.span_from_boundary(self.inner.word_boundaries[index]),
         }
     }
 
-    fn err_through_end_from(&self, index: usize, err_msg: &'static str) -> ParseError {
+    fn err_through_end_from(&self, index: usize, err_msg: impl Into<String>) -> ParseError {
         ParseError {
-            kind: ParseErrorKind::InvalidCommand(err_msg),
+            kind: ParseErrorKind::InvalidCommand(err_msg.into()),
             span: self.span_through_end(index),
         }
     }
 
+    fn ensure_zero_args(
+        &mut self,
+        cmd: &'static str,
+        err_msg_suffix: &'static str,
+    ) -> Result<(), ParseError> {
+        let message = format!("`{}` takes 0 arguments{}", cmd, err_msg_suffix);
+        if self.inner.words.next().is_some() {
+            return Err(self.err_through_end_from(1, message));
+        }
+        Ok(())
+    }
+
     /// Attempt to parse the input provided to Self::new as a command, such as "help",
     /// "list <args>", "quit", etc
     ///
@@ -159,20 +171,6 @@ impl<'a> CommandParser<'a> {
     /// - `Err(ParseError)`, if the input starts with a valid command but has the wrong
     ///   number or kind of arguments, e.g. `list foobar`
     pub fn parse_command(&mut self) -> Result<Command, ParseError> {
-        macro_rules! ensure_zero_args {
-            ($parser:expr, $cmd:literal) => {{
-                ensure_zero_args!($parser, $cmd, "")
-            }};
-            ($parser:expr, $cmd: literal, $err_msg_suffix:literal) => {{
-                if $parser.inner.words.next().is_some() {
-                    return Err($parser.err_through_end_from(
-                        1,
-                        concat!("`", $cmd, "` takes 0 arguments", $err_msg_suffix),
-                    ));
-                }
-            }};
-        }
-
         // expect() is guaranteed to succeed; CommandParser::new ensures there is at
         // least one word
         let command = match self
@@ -182,23 +180,19 @@ impl<'a> CommandParser<'a> {
             .expect("CommandParser::new is supposed to check that we have a valid command")
         {
             "help" => {
-                ensure_zero_args!(
-                    self,
-                    "help",
-                    "; use `info <item>` for information about an item"
-                );
+                self.ensure_zero_args("help", "; use `info <item>` for information about an item")?;
                 Command::Help
             }
             "clear" => {
-                ensure_zero_args!(self, "clear");
+                self.ensure_zero_args("clear", "")?;
                 Command::Clear
             }
             "quit" => {
-                ensure_zero_args!(self, "quit");
+                self.ensure_zero_args("quit", "")?;
                 Command::Quit
             }
             "exit" => {
-                ensure_zero_args!(self, "exit");
+                self.ensure_zero_args("exit", "")?;
                 Command::Quit
             }
             "info" => {
diff --git a/numbat/src/parser.rs b/numbat/src/parser.rs
index 144473c..9067121 100644
--- a/numbat/src/parser.rs
+++ b/numbat/src/parser.rs
@@ -238,7 +238,7 @@ pub enum ParseErrorKind {
     ExpectedLocalVariableDefinition,
 
     #[error("Invalid command: {0}")]
-    InvalidCommand(&'static str),
+    InvalidCommand(String),
 }
 
 #[derive(Debug, Clone, Error)]
```

Reply:
If you're ok with those allocations, then sure.